File: electron/tests/capabilities_test.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

import { describe, expect, it } from "vitest";

import {
  ALL_CAPABILITIES,
  capabilityList,
  isStamped,
  NO_CAPABILITIES,
  readBuildCapabilities,
  resolveCapabilities,
} from "../src/capabilities";

const NOTHING = NO_CAPABILITIES;

describe("the package stamp", () => {
  it("gives an unstamped build nothing — the same as a download", () => {
    expect(readBuildCapabilities({})).toEqual(NO_CAPABILITIES);
    expect(readBuildCapabilities(null)).toEqual(NO_CAPABILITIES);
    expect(readBuildCapabilities({ capabilities: "yes" })).toEqual(
      NO_CAPABILITIES,
    );
  });

  it("reads each capability off the manifest", () => {
    expect(
      readBuildCapabilities({
        capabilities: { multiplayer: true, mods: true, licensed: true },
      }),
    ).toEqual({
      multiplayer: true,
      mods: true,
      portMap: false,
      voice: false,
      licensed: true,
    });
  });

  it("treats an absent field as off rather than on", () => {
    expect(readBuildCapabilities({ capabilities: { mods: true } })).toEqual({
      ...NO_CAPABILITIES,
      mods: true,
    });
  });
});

describe("telling a developer build apart", () => {
  it("counts anything nobody packaged as one", () => {
    expect(isStamped({})).toBe(false);
    expect(isStamped(null)).toBe(false);
    expect(isStamped({ capabilities: "yes" })).toBe(false);
  });

  it("does not count a packaged build, however narrow its stamp", () => {
    expect(isStamped({ capabilities: NOTHING })).toBe(true);
    expect(isStamped({ capabilities: ALL_CAPABILITIES })).toBe(true);
  });
});

describe("what a launch may do", () => {
  it("leaves a stamped build alone when nothing is asked for", () => {
    const { capabilities } = resolveCapabilities(NOTHING, ["game"]);
    expect(capabilities.multiplayer).toBe(false);
    expect(capabilities.mods).toBe(false);
    expect(capabilities.unlocked).toBe(false);
  });

  it("unlocks each feature on its own word", () => {
    const { capabilities, refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--multiplayer",
      "--port",
      "27849",
      "--mods",
    ]);
    expect(capabilities).toMatchObject({
      multiplayer: true,
      mods: true,
      portMap: false,
      direct: true,
      port: 27849,
      unlocked: true,
    });
    expect(refusals).toEqual([]);
  });

  it("accepts the --port=N spelling", () => {
    const { capabilities } = resolveCapabilities(NOTHING, [
      "game",
      "--multiplayer",
      "--port=27015",
    ]);
    expect(capabilities.port).toBe(27015);
  });

  it("takes --multiplayer with no port — the HOST screen has one", () => {
    const { capabilities, refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--multiplayer",
    ]);
    expect(capabilities.multiplayer).toBe(true);
    expect(capabilities.direct).toBe(false);
    expect(capabilities.port).toBeUndefined();
    expect(refusals).toEqual([]);
  });

  it("refuses a port that is not one", () => {
    const { capabilities, refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--multiplayer",
      "--port",
      "no",
    ]);
    expect(capabilities.port).toBeUndefined();
    expect(refusals).toEqual(["--port no is not a port number"]);
  });

  it("says so when a door was asked for with no session to put it on", () => {
    const { refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--port",
      "27849",
    ]);
    expect(refusals).toEqual(["--port does nothing without --multiplayer"]);
  });

  it("never turns the router mapping on from the command line", () => {
    const { capabilities } = resolveCapabilities(NOTHING, [
      "game",
      "--upnp",
      "--portmap",
      "--multiplayer",
      "--port",
      "27849",
    ]);
    expect(capabilities.portMap).toBe(false);
  });

  it("does not call a build that already had them unlocked", () => {
    const { capabilities } = resolveCapabilities(ALL_CAPABILITIES, [
      "game",
      "--mods",
    ]);
    expect(capabilities.unlocked).toBe(false);
  });

  it("takes the licence as a word, on either the stamp or the launch", () => {
    expect(resolveCapabilities(NOTHING, ["game"]).capabilities.licensed).toBe(
      false,
    );
    expect(
      resolveCapabilities(NOTHING, ["game", "--licensed"]).capabilities
        .licensed,
    ).toBe(true);
    expect(
      resolveCapabilities({ ...NOTHING, licensed: true }, ["game"]).capabilities
        .licensed,
    ).toBe(true);
  });

  it("does not treat the licence claim as an unlock of its own", () => {
    // It opens no feature — it says a held licence covers the ones that are
    // already open — so on its own it raises nothing to acknowledge.
    const { capabilities } = resolveCapabilities(NOTHING, [
      "game",
      "--licensed",
    ]);
    expect(capabilities.multiplayer).toBe(false);
    expect(capabilities.unlocked).toBe(false);
  });

  it("leaves a depot build's own doors alone", () => {
    const { capabilities } = resolveCapabilities(ALL_CAPABILITIES, ["game"]);
    expect(capabilities).toMatchObject({
      multiplayer: true,
      mods: true,
      portMap: true,
      licensed: true,
      direct: false,
      unlocked: false,
    });
  });

  it("hands the page only the names it may act on", () => {
    expect(
      capabilityList({
        ...NOTHING,
        autopilot: false,
        unlocked: false,
        direct: false,
      }),
    ).toEqual([]);
    expect(
      capabilityList({
        ...ALL_CAPABILITIES,
        autopilot: false,
        unlocked: false,
        direct: false,
      }),
    ).toEqual(["multiplayer", "mods", "voice"]);
  });

  // THE AUTO PILOT — the one capability no desktop build carries, depot build
  // included. It is not in `ALL_CAPABILITIES` because there is nothing for a
  // packager to stamp: `--autopilot` is the only way it is ever on, and it
  // costs the launch its multiplayer to ask.
  it("gives the depot build no auto pilot to offer", () => {
    const { capabilities } = resolveCapabilities(ALL_CAPABILITIES, ["game"]);
    expect(capabilities.autopilot).toBe(false);
    expect(capabilityList(capabilities)).not.toContain("autopilot");
  });

  it("hands the page the auto pilot when the command line asked for it", () => {
    const { capabilities } = resolveCapabilities(NOTHING, [
      "game",
      "--autopilot",
    ]);
    expect(capabilities.autopilot).toBe(true);
    expect(capabilityList(capabilities)).toContain("autopilot");
  });

  it("takes a depot build's multiplayer away to give it the auto pilot", () => {
    // The stamp is not a defence: a copy that plays itself has no business in
    // somebody else's session whatever packaged it.
    const { capabilities, refusals } = resolveCapabilities(ALL_CAPABILITIES, [
      "game",
      "--autopilot",
    ]);
    expect(capabilities).toMatchObject({
      autopilot: true,
      multiplayer: false,
      voice: false,
      licensed: false,
    });
    expect(capabilityList(capabilities)).toEqual(["mods", "autopilot"]);
    expect(refusals).toEqual([
      "--autopilot is a developer switch: multiplayer is off for this launch",
    ]);
  });

  it("takes it away from the command line too, in either order", () => {
    for (const argv of [
      ["game", "--multiplayer", "--voice", "--licensed", "--autopilot"],
      ["game", "--autopilot", "--multiplayer", "--voice", "--licensed"],
    ]) {
      const { capabilities } = resolveCapabilities(NOTHING, argv);
      expect(capabilities, argv.join(" ")).toMatchObject({
        autopilot: true,
        multiplayer: false,
        voice: false,
        licensed: false,
      });
    }
  });

  it("states the consequence once, and blames no flag for it", () => {
    // The pairing refusals below would otherwise fire on every autopilot
    // launch, naming flags the developer did not type — which reads as a bug
    // in the parser rather than as the one deliberate trade.
    const { refusals } = resolveCapabilities(ALL_CAPABILITIES, [
      "game",
      "--autopilot",
      "--port",
      "27849",
    ]);
    expect(refusals).toEqual([
      "--autopilot is a developer switch: multiplayer is off for this launch",
    ]);
  });

  it("does not call the auto pilot an unlock of its own", () => {
    // `unlocked` is the LICENCE acknowledgement — it is about running licensed
    // features outside the terms they came under. The auto pilot has its own
    // fact and its own paragraph in the same notice.
    const { capabilities } = resolveCapabilities(NOTHING, [
      "game",
      "--autopilot",
    ]);
    expect(capabilities.unlocked).toBe(false);
  });

  // VOICE — a capability of its own, because it opens a microphone and makes the
  // host relay every speaker to every listener. The depot build carries it; a
  // plain download does not.
  it("keeps voice off a build that was not given it", () => {
    const { capabilities } = resolveCapabilities(NOTHING, ["game"]);
    expect(capabilities.voice).toBe(false);
    expect(capabilityList(capabilities)).not.toContain("voice");
  });

  it("lets the command line turn voice on for one launch, and says so", () => {
    const { capabilities, refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--multiplayer",
      "--voice",
    ]);
    expect(capabilities.voice).toBe(true);
    // `unlocked` is what makes the launch print the notice: a capability the
    // package did not carry was turned on by hand.
    expect(capabilities.unlocked).toBe(true);
    expect(refusals).toEqual([]);
  });

  // THE PAIRING. Voice travels inside a session (`FRAME.voice`, relayed by the
  // session server), so on a build that can neither host nor join one there is
  // nothing for a microphone to talk into — and granting the device anyway
  // would put a VOICE CHAT page in the settings that never carries a syllable.
  it("refuses voice without multiplayer, by name rather than silently", () => {
    const { capabilities, refusals } = resolveCapabilities(NOTHING, [
      "game",
      "--voice",
    ]);
    expect(capabilities.voice).toBe(false);
    expect(refusals).toEqual(["--voice does nothing without --multiplayer"]);
  });

  it("refuses a STAMPED voice build that carries no multiplayer", () => {
    // The same rule from the other direction: a packaging run that enabled
    // voice and forgot multiplayer must not produce a build with an open
    // microphone and nowhere to send it.
    const { capabilities } = resolveCapabilities({ ...NOTHING, voice: true }, [
      "game",
    ]);
    expect(capabilities.voice).toBe(false);
  });
});

File: tauri/shell/tests/capabilities_test.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! What a build may do, and what a command line may add to it — the peer of
//! `electron/tests/capabilities_test.ts`.
//!
//! The rule under every case here is the same one: a build carries only what
//! something DELIBERATELY gave it. The narrow answer is what happens by
//! default, and the wide one has to be asked for — so that a tree somebody
//! cloned behaves like the thing they would have downloaded.

use adastrail_shell::capabilities::{
    capability_list, is_stamped, read_build_capabilities, resolve_capabilities, BuildCapabilities,
    BuildStamp, ALL_CAPABILITIES, NO_CAPABILITIES,
};

fn argv(args: &[&str]) -> Vec<String> {
    args.iter().map(|arg| (*arg).to_string()).collect()
}

const DEPOT: BuildStamp<'static> = BuildStamp {
    stamped: Some("1"),
    multiplayer: Some("1"),
    mods: Some("1"),
    port_map: Some("1"),
    voice: Some("1"),
    licensed: Some("1"),
};

#[test]
fn an_unstamped_build_carries_nothing() {
    let stamp = BuildStamp::default();
    assert_eq!(read_build_capabilities(&stamp), NO_CAPABILITIES);
    assert!(!is_stamped(&stamp), "a checkout is a developer build");
}

#[test]
fn a_stamp_that_says_nothing_still_carries_nothing() {
    // The stamp is present (this was packaged) but every switch was off — a
    // plain download.
    let stamp = BuildStamp {
        stamped: Some("1"),
        ..BuildStamp::default()
    };
    assert!(is_stamped(&stamp), "packaged, just not with anything on");
    assert_eq!(read_build_capabilities(&stamp), NO_CAPABILITIES);
}

#[test]
fn the_depot_stamp_carries_everything() {
    assert_eq!(read_build_capabilities(&DEPOT), ALL_CAPABILITIES);
    assert!(is_stamped(&DEPOT));
}

#[test]
fn a_switch_set_on_an_unstamped_build_is_still_nothing() {
    // Somebody's shell has GIS_ENABLE_MULTIPLAYER lying around from a packaging
    // run; a `cargo build` after that must not quietly produce a wide binary.
    let stamp = BuildStamp {
        multiplayer: Some("1"),
        ..BuildStamp::default()
    };
    assert_eq!(read_build_capabilities(&stamp), NO_CAPABILITIES);
}

#[test]
fn only_an_explicit_yes_turns_a_switch_on() {
    let stamp = BuildStamp {
        stamped: Some("1"),
        multiplayer: Some("0"),
        mods: Some(""),
        voice: Some("yes"),
        ..BuildStamp::default()
    };
    assert_eq!(read_build_capabilities(&stamp), NO_CAPABILITIES);
}

#[test]
fn the_command_line_can_widen_a_download_and_says_so() {
    let (caps, refusals) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--multiplayer"]));
    assert!(caps.multiplayer());
    assert!(caps.unlocked, "the launch has to announce this");
    assert!(refusals.is_empty());
}

#[test]
fn a_depot_build_is_not_unlocked_by_asking_for_what_it_has() {
    let (caps, refusals) = resolve_capabilities(ALL_CAPABILITIES, &argv(&["--multiplayer"]));
    assert!(caps.multiplayer());
    assert!(!caps.unlocked, "it already had it — nothing to announce");
    assert!(refusals.is_empty());
}

#[test]
fn voice_without_a_session_is_refused_by_name() {
    // Not ignored: "--voice does nothing" would leave somebody adding it a
    // second time and louder. And it must not grant the microphone.
    let (caps, refusals) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--voice"]));
    assert!(!caps.voice());
    assert_eq!(refusals, vec!["--voice does nothing without --multiplayer"]);
}

#[test]
fn voice_with_a_session_is_granted() {
    let (caps, refusals) =
        resolve_capabilities(NO_CAPABILITIES, &argv(&["--multiplayer", "--voice"]));
    assert!(caps.voice());
    assert!(caps.unlocked);
    assert!(refusals.is_empty());
}

#[test]
fn the_router_permission_cannot_be_asked_for() {
    // A port mapping is a change this program makes to somebody's ROUTER, so it
    // belongs to the build and to nothing else.
    let (caps, _) = resolve_capabilities(
        NO_CAPABILITIES,
        &argv(&["--multiplayer", "--portmap", "--upnp"]),
    );
    assert!(!caps.port_map());

    let built = BuildCapabilities {
        port_map: true,
        ..NO_CAPABILITIES
    };
    let (caps, _) = resolve_capabilities(built, &argv(&[]));
    assert!(caps.port_map());
}

#[test]
fn a_port_pins_the_direct_door_both_ways_round() {
    for args in [
        argv(&["--multiplayer", "--port", "27849"]),
        argv(&["--multiplayer", "--port=27849"]),
    ] {
        let (caps, refusals) = resolve_capabilities(NO_CAPABILITIES, &args);
        assert!(caps.direct);
        assert_eq!(caps.port, Some(27849));
        assert!(refusals.is_empty());
    }
}

#[test]
fn a_port_without_a_session_is_refused() {
    let (caps, refusals) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--port", "27849"]));
    assert!(!caps.direct);
    assert_eq!(caps.port, None);
    assert_eq!(refusals, vec!["--port does nothing without --multiplayer"]);
}

#[test]
fn a_port_that_is_not_a_port_is_refused_by_value() {
    for bad in ["0", "70000", "-1", "twenty", ""] {
        let (caps, refusals) = resolve_capabilities(ALL_CAPABILITIES, &argv(&["--port", bad]));
        assert!(!caps.direct, "{bad} must not open a door");
        assert_eq!(refusals.len(), 1, "{bad} must be refused by name");
        assert!(refusals[0].contains(bad) || bad.is_empty());
    }
}

#[test]
fn the_licence_is_a_declaration_rather_than_an_unlock() {
    // Nothing here can check it and nothing pretends to — but claiming it is
    // not "unlocking" anything either, so it does not trip the announcement.
    let (caps, _) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--licensed"]));
    assert!(caps.licensed());
    assert!(!caps.unlocked);
}

#[test]
fn the_page_is_told_only_plain_names() {
    let (caps, _) = resolve_capabilities(ALL_CAPABILITIES, &argv(&[]));
    assert_eq!(capability_list(&caps), vec!["multiplayer", "mods", "voice"]);

    let (caps, _) = resolve_capabilities(NO_CAPABILITIES, &argv(&[]));
    assert!(
        capability_list(&caps).is_empty(),
        "a download offers the page nothing to draw a door for"
    );
}

// THE AUTO PILOT — the one capability no desktop build carries, depot build
// included. It is not in `ALL_CAPABILITIES` because there is nothing for a
// packager to stamp: `--autopilot` is the only way it is ever on, and it costs
// the launch its multiplayer to ask.

#[test]
fn no_build_stamp_can_hand_out_the_auto_pilot() {
    let (caps, _) = resolve_capabilities(ALL_CAPABILITIES, &argv(&[]));
    assert!(!caps.autopilot());
    assert!(!capability_list(&caps).contains(&"autopilot"));
}

#[test]
fn the_command_line_hands_the_page_the_auto_pilot() {
    let (caps, _) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--autopilot"]));
    assert!(caps.autopilot());
    assert!(capability_list(&caps).contains(&"autopilot"));
}

#[test]
fn the_auto_pilot_costs_the_launch_its_multiplayer() {
    // The stamp is not a defence: a copy that plays itself has no business in
    // somebody else's session whatever built it.
    let (caps, refusals) = resolve_capabilities(ALL_CAPABILITIES, &argv(&["--autopilot"]));
    assert!(caps.autopilot());
    assert!(!caps.multiplayer());
    assert!(!caps.voice());
    assert!(!caps.licensed());
    assert_eq!(capability_list(&caps), vec!["mods", "autopilot"]);
    assert_eq!(
        refusals,
        vec!["--autopilot is a developer switch: multiplayer is off for this launch"]
    );
}

#[test]
fn the_auto_pilot_beats_the_command_line_too_in_either_order() {
    for args in [
        ["--multiplayer", "--voice", "--licensed", "--autopilot"],
        ["--autopilot", "--multiplayer", "--voice", "--licensed"],
    ] {
        let (caps, _) = resolve_capabilities(NO_CAPABILITIES, &argv(&args));
        assert!(caps.autopilot(), "{args:?}");
        assert!(!caps.multiplayer(), "{args:?}");
        assert!(!caps.voice(), "{args:?}");
        assert!(!caps.licensed(), "{args:?}");
    }
}

#[test]
fn the_consequence_is_stated_once_and_blames_no_flag_for_it() {
    // The pairing refusals would otherwise fire on every autopilot launch,
    // naming flags the developer did not type — which reads as a bug in the
    // parser rather than as the one deliberate trade.
    let (_, refusals) =
        resolve_capabilities(ALL_CAPABILITIES, &argv(&["--autopilot", "--port", "27849"]));
    assert_eq!(
        refusals,
        vec!["--autopilot is a developer switch: multiplayer is off for this launch"]
    );
}

#[test]
fn the_auto_pilot_is_not_a_licence_unlock_of_its_own() {
    // `unlocked` is the LICENCE acknowledgement — about running licensed
    // features outside the terms they came under. The ride has its own fact and
    // its own paragraph in the same notice.
    let (caps, _) = resolve_capabilities(NO_CAPABILITIES, &argv(&["--autopilot"]));
    assert!(!caps.unlocked);
}
